component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,272 @@
+        Map<String, String[]> requestParams = Arrays.stream(params).collect(Collectors.toMap(RequestParameter::getKey,
+                RequestParameter::getValue));
+        if (requestParams.get(TokenExchangeConstants.REQUESTED_TOKEN_TYPE) != null) {
+            requested_token_type = requestParams.get(TokenExchangeConstants.REQUESTED_TOKEN_TYPE)[0];
+        }
+        if(requestParams.get(TokenExchangeConstants.AUDIENCE) != null) {
+            requested_audience = requestParams.get(TokenExchangeConstants.AUDIENCE)[0];
+        }
+        String tenantDomain = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain();
+        if (StringUtils.isEmpty(tenantDomain)) {
+            tenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+        }
+        if (TokenExchangeConstants.JWT_TOKEN_TYPE.equals(requested_token_type)) {
+            signedJWT = TokenExchangeUtils.getSignedJWT(requestParams.get(TokenExchangeConstants.SUBJECT_TOKEN)[0]);
+            if (signedJWT == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "No Valid subject token was found for "
+                        + TokenExchangeConstants.TOKEN_EXCHANGE_GRANT_TYPE);
+            } else {
+                claimsSet = TokenExchangeUtils.getClaimSet(signedJWT);
+            }
+
+            if (claimsSet == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Claim values are empty in the given JSON Web Token");
+            }
+
+            String jwtIssuer = claimsSet.getIssuer();
+            String subject = resolveSubject(claimsSet);
+            List<String> audiences = claimsSet.getAudience();
+            Date expirationTime = claimsSet.getExpirationTime();
+            Date notBeforeTime = claimsSet.getNotBeforeTime();
+            Date issuedAtTime = claimsSet.getIssueTime();
+            Map<String, Object> customClaims = new HashMap<>(claimsSet.getClaims());
+
+            tokReqMsgCtx.addProperty(TokenExchangeConstants.EXPIRY_TIME, expirationTime);
+            long currentTimeInMillis = System.currentTimeMillis();
+            long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
+
+            if (StringUtils.isEmpty(jwtIssuer) || StringUtils.isEmpty(subject) || expirationTime == null ||
+                    audiences == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Mandatory fields(Issuer, Subject, Expiration time "
+                        + "or Audience) are empty in the given JWT");
+            }
+            identityProvider = TokenExchangeUtils.getIdPByIssuer(jwtIssuer, tenantDomain);
+            tokenEndPointAlias = TokenExchangeUtils.getTokenEndpointAlias(identityProvider, tenantDomain);
+            try {
+                if (signedJWT != null) {
+                    if (TokenExchangeUtils.validateSignature(signedJWT, identityProvider, tenantDomain)) {
```

Comment:
can use this https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/JWTSignatureValidationUtils.java#L67 to validate signature

Reply:
Same as https://github.com/wso2-extensions/identity-oauth2-grant-token-exchange/pull/1#discussion_r650674000